### PR TITLE
Report "Build ID" as codeIdentifier in NDK stacktraces

### PR DESCRIPTION
**Pull request by lemnik**
## Goal
When reporting native stack traces, we need to include the "Build ID" as `codeIdentifier` for each stack frame (when it's available) so that we can later retrieve the correct symbol data.

## Changeset

- Add the "Build ID" to the stack-frames in the NDK event model during a crash
- Use `NativeStackframe` instead of `StackTraceElement` to relay `bugsnag_notify` events to the JVM layer

## Testing
New unit tests and additional checks on top of existing end-to-end testing

### CODE REVIEW

File: bugsnag-android-core/src/main/java/com/bugsnag/android/NativeInterface.java [hunk 1]
```diff
@@ -38,6 +42,36 @@ private static Client getClient() {
         }
     }
 
+    /**
+     * Create an empty Event for a native "handled exception" report. The returned Event will have
+     * no Error objects, metadata, breadcrumbs, or feature flags. It's indended that the caller
```
Comment (kattrali) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/NativeInterface.java:47]:
```suggestion
     * no Error objects, metadata, breadcrumbs, or feature flags. It's intended that the caller
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/NativeInterface.java [hunk 2]
```diff
@@ -38,6 +42,36 @@ private static Client getClient() {
...
+     * will populate the Error and then pass the Event object to
+     * {@link Client#populateAndNotifyAndroidEvent(Event, OnErrorCallback)}.
+     */
+    private static Event createNativeEvent() {
```
Comment (kattrali) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/NativeInterface.java:51]: What about this makes it inherently a native event? Does `createEmptyEvent()` make sense here (or perhaps as a convenience initializer on Event)?

File: bugsnag-android-core/src/test/java/com/bugsnag/android/NativeInterfaceApiTest.kt [hunk 3]
```diff
@@ -220,11 +235,22 @@ internal class NativeInterfaceApiTest {
...
+            Severity.ERROR,
+            arrayOf<StackTraceElement>()
+        )
         verify(client, times(1)).notify(any(), any())
     }
 
+    @Test
+    fun notifyNativeTraceCall() {
+        NativeInterface.notify("SIGPIPE", "SIGSEGV 11", Severity.ERROR, arrayOf<NativeStackframe>())
+        verify(client, times(1)).populateAndNotifyAndroidEvent(any(), any())
```
Comment (kattrali) [inline bugsnag-android-core/src/test/java/com/bugsnag/android/NativeInterfaceApiTest.kt:251]: perhaps there's something more specific to be verified here, like to ensure the event was created with the correct error class / message?

File: bugsnag-plugin-android-ndk/src/main/jni/utils/string.c [hunk 4]
```diff
@@ -24,3 +24,20 @@ void bsg_strncpy(char *dst, const char *src, size_t dst_size) {
...
+      (max_chars > byte_count * 2) ? byte_count : (max_chars - 1) / 2;
+
+  char *cursor = (char *)src;
+  char *outCursor = dst;
+  for (size_t i = 0; i < byte_copy_count; ++i) {
+    *outCursor++ = hex[(*cursor >> 4) & 0xF];
+    *outCursor++ = hex[(*cursor++) & 0xF];
+  }
+
+  *outCursor = 0;
```
Comment (kattrali) [inline bugsnag-plugin-android-ndk/src/main/jni/utils/string.c:42]: What is this set again for?

Comment (lemnik) [reply bugsnag-plugin-android-ndk/src/main/jni/utils/string.c:42]: null-terminator, maybe `\0` would be clearer?

File: bugsnag-android-core/src/main/java/com/bugsnag/android/NativeInterface.java (hunk 2 above)
Comment (lemnik) [reply bugsnag-android-core/src/main/java/com/bugsnag/android/NativeInterface.java:51]: Actually I can use an existing constructor and get the same effect.

**Pull request merged**